### PR TITLE
Adding ExternalTestServer class

### PR DESCRIPTION
**Pull request by mosquito**
## Adding class `ExternalTestServer`

In our projects application running in separate task. And we have to set `_root` property for our test cases.
This improvement adding this opportunity as a special class.

## Are there changes in behaviour for the user?

No

## Related issue number

No one issue.

## Checklist

- [X] I think the code is well written
- [X] Unit tests for the changes exist
- [ ] Documentation reflects the changes
- [X] If you provide code modification, please add yourself to `CONTRIBUTORS.txt`
  * The format is &lt;Name&gt; &lt;Surname&gt;.
  * Please keep alphabetical order, the file is sorted by names. 
- [x] Add a new news fragment into the `CHANGES` folder
  * name it `<issue_id>.<type>` for example (588.bugfix)
  * if you don't have an `issue_id` change it to the pr id after creating the pr
  * ensure type is one of the following:
    * `.feature`: Signifying a new feature.
    * `.bugfix`: Signifying a bug fix.
    * `.doc`: Signifying a documentation improvement.
    * `.removal`: Signifying a deprecation or removal of public API.
    * `.misc`: A ticket has been closed, but it is not of interest to users.
  * Make sure to use full sentences with correct case and punctuation, for example: "Fix issue with non-ascii contents in doctest text files."

### CODE REVIEW

**Comment by prostomarkeloff** [comment]
Looks useful for me.

**Comment by mosquito** [comment]
@asvetlov done.

File: aiohttp/test_utils.py [hunk 1]
```diff
@@ -74,26 +70,75 @@ def unused_port() -> int:
...
+
+    def __init__(self, root: URL, *,
+                 skip_url_asserts: bool = False,
+                 **kwargs: Any):
+        assert root.is_absolute(), "Only absolute URL allowed"
+        super().__init__(skip_url_asserts=skip_url_asserts, **kwargs)
+        self._root = root       # type: URL
+
+    async def start_server(self) -> None:
+        raise NotImplementedError(
```
Comment (asvetlov) [inline aiohttp/test_utils.py]: I think better to do nothing here instead of raising an exception

File: aiohttp/test_utils.py [hunk 2]
```diff
@@ -74,26 +70,75 @@ def unused_port() -> int:
...
+            "ExternalTestServer doesn't support start_server"
+        )
+
+    async def close(self) -> None:
+        return
+
+    @property
+    def host(self) -> str:
```
Comment (asvetlov) [inline aiohttp/test_utils.py]: Should these properties be a part of base class? I think so.

File: aiohttp/test_utils.py [hunk 3]
```diff
@@ -74,51 +70,110 @@ def unused_port() -> int:
...
+    @property
+    def scheme(self) -> str:
+        return self._root.scheme
+
+
+class BaseTestServer(AbstractTestServer):
+    __test__ = False
+
+    def __init__(self, *,
+                 scheme: Optional[str] = None,
```
Comment (mosquito) [inline aiohttp/test_utils.py:172]: Did not found any reason for using `sentinel` instead of `None`.

File: aiohttp/test_utils.py [hunk 4]
```diff
@@ -74,51 +70,110 @@ def unused_port() -> int:
...
         if self.runner:
             return
         self._ssl = kwargs.pop('ssl', None)
         self.runner = await self._make_runner(**kwargs)
         await self.runner.setup()
-        if not self.port:
-            self.port = 0
-        _sock = get_port_socket(self.host, self.port)
-        self.host, self.port = _sock.getsockname()[:2]
+        _sock = get_port_socket(self.host, self.port or 0)
```
Comment (mosquito) [inline aiohttp/test_utils.py]: prevent mypy warnings

File: aiohttp/test_utils.py [hunk 5]
```diff
@@ -74,51 +70,110 @@ def unused_port() -> int:
...
-            else:
-                scheme = 'http'
-            self.scheme = scheme
-        self._root = URL('{}://{}:{}'.format(self.scheme,
-                                             self.host,
-                                             self.port))
+        self._port = sockets[0].getsockname()[1]
+
+        if self.scheme is None:
+            self._scheme = 'https' if self._ssl else 'http'
```
Comment (mosquito) [inline aiohttp/test_utils.py]: shorter and much simpler

File: aiohttp/test_utils.py (hunk 2 above)
Comment (asvetlov) [reply aiohttp/test_utils.py]: Could it return `self._root.host`?

Comment (asvetlov) [reply aiohttp/test_utils.py]: The same for other properties.

**Review by asvetlov — APPROVED** [review]

**Review by asvetlov — CHANGES_REQUESTED** [review]
Sorry, looks like the PR is broken; tests fail on the current master.

Please fix them.
Otherwise, I have to close the PR.

Sorry again, I should review faster to prevent such situations.

**Comment by mosquito** [comment]
It is still needed, but there are too many conflicts in it and I just close it. Sadly, I still need to change the underscored property in our projects.

**Comment by asvetlov** [comment]
Please feel free to publish an updated PR version

**Pull request closed without merging**